File: src/frontend/src/types/problem.ts
export interface Problem {
  id: number;
  leetcode_id: number | null;
  title: string;
  url: string | null;
  difficulty: "easy" | "medium" | "hard" | null;
  tags: string[];
  pattern: string | null;
  category: "algorithm" | "ml_coding" | "system_design";
  source: string | null;
  company_tags: string[];
  priority: number;
  is_completed: boolean;
  comfort_level: number;
  created_at: string;
  last_attempted_at: string | null;
  next_review_at: string | null;
  framework_node_id: number | null;
  description: string | null;
  neetcode_slug: string | null;
  description_source: string | null;
  notes: string | null;
}

export type Difficulty = "easy" | "medium" | "hard";
export type Category = "algorithm" | "ml_coding" | "system_design";
export type SortField =
  | "comfort_level"
  | "last_attempted_at"
  | "next_review_at"
  | "created_at"
  | "difficulty";
export type SortOrder = "asc" | "desc";

export type AttemptResult = "solved" | "hint" | "failed" | "timeout";

export interface AttemptCreate {
  duration_seconds: number | null;
  result: AttemptResult;
  approach_notes: string | null;
  complexity_time: string | null;
  complexity_space: string | null;
  comfort_after: number;
}

export interface Attempt {
  id: number;
  problem_id: number;
  started_at: string | null;
  duration_seconds: number | null;
  result: string | null;
  approach_notes: string | null;
  complexity_time: string | null;
  complexity_space: string | null;
  llm_review: string | null;
  comfort_after: number | null;
}

export interface ProblemFilters {
  difficulty?: Difficulty;
  pattern?: string;
  source?: string;
  company?: string;
  is_completed?: boolean;
  category?: Category;
  sort_by: SortField;
  sort_order: SortOrder;
  limit: number;
  offset: number;
}

// --- Stats types ---

export interface PatternStat {
  pattern: string;
  count: number;
  completed: number;
  avg_comfort: number;
}

export interface ProblemStats {
  total: number;
  completed: number;
  avg_comfort: number;
  by_difficulty: Record<string, number>;
  by_pattern: PatternStat[];
  weak_patterns: string[];
  total_attempts: number;
  avg_duration_seconds: number;
}

// --- Review types ---

export type ReviewVerdict =
  | "optimal"
  | "suboptimal"
  | "incorrect"
  | "needs_clarification";

export interface ReviewResult {
  verdict: ReviewVerdict;
  feedback: string;
  hint: string | null;
  optimal_complexity: { time: string; space: string };
  pattern: string;
  follow_up: string;
}

// --- QA types ---

export interface QAChatMessage {
  role: "user" | "assistant";
  content: string;
  timestamp: string;
}

export interface QAChatResponse {
  session_id: number;
  reply: string;
  messages: QAChatMessage[];
}

export interface QASessionSummary {
  id: number;
  problem_id: number | null;
  topic: string | null;
  created_at: string | null;
  summary: string | null;
}

File: src/backend/schemas/problem.py
"""Pydantic schemas for Problem and Attempt CRUD."""
import json
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


class ProblemCreate(BaseModel):
    """Schema for creating a new problem."""

    leetcode_id: int | None = None
    title: str = Field(min_length=1)
    url: str | None = None
    difficulty: Literal["easy", "medium", "hard"] | None = None
    tags: list[str] = []
    pattern: str | None = None
    category: Literal["algorithm", "ml_coding", "system_design"] = "algorithm"
    source: str | None = None
    company_tags: list[str] = []
    priority: int = Field(default=2, ge=1, le=3)
    framework_node_id: int | None = None
    description: str | None = None
    neetcode_slug: str | None = None
    description_source: str | None = None
    notes: str | None = None


class ProblemUpdate(BaseModel):
    """Schema for partial problem update."""

    leetcode_id: int | None = None
    title: str | None = Field(default=None, min_length=1)
    url: str | None = None
    difficulty: Literal["easy", "medium", "hard"] | None = None
    tags: list[str] | None = None
    pattern: str | None = None
    category: Literal["algorithm", "ml_coding", "system_design"] | None = None
    source: str | None = None
    company_tags: list[str] | None = None
    priority: int | None = Field(default=None, ge=1, le=3)
    is_completed: bool | None = None
    comfort_level: int | None = Field(default=None, ge=0, le=5)
    framework_node_id: int | None = None
    description: str | None = None
    neetcode_slug: str | None = None
    description_source: str | None = None
    notes: str | None = None


class ProblemResponse(BaseModel):
    """Schema for problem API response."""

    id: int
    leetcode_id: int | None = None
    title: str
    url: str | None = None
    difficulty: str | None = None
    tags: list[str] = []
    pattern: str | None = None
    category: str = "algorithm"
    source: str | None = None
    company_tags: list[str] = []
    priority: int | None = 2
    is_completed: bool = False
    comfort_level: int = 0
    created_at: datetime | None = None
    last_attempted_at: datetime | None = None
    next_review_at: datetime | None = None
    framework_node_id: int | None = None
    description: str | None = None
    neetcode_slug: str | None = None
    description_source: str | None = None
    notes: str | None = None

    model_config = ConfigDict(from_attributes=True)

    @field_validator("tags", "company_tags", mode="before")
    @classmethod
    def parse_json_list(cls, v: object) -> list[str]:
        """Parse JSON string to list if needed (SQLAlchemy stores as text)."""
        if v is None:
            return []
        if isinstance(v, str):
            return json.loads(v)
        return v  # type: ignore[return-value]


class AttemptCreate(BaseModel):
    """Schema for creating a new attempt."""

    duration_seconds: int | None = Field(default=None, ge=0)
    result: Literal["solved", "hint", "failed", "timeout"]
    approach_notes: str | None = None
    complexity_time: str | None = None
    complexity_space: str | None = None
    comfort_after: int = Field(ge=1, le=5)


class AttemptResponse(BaseModel):
    """Schema for attempt API response."""

    id: int
    problem_id: int
    started_at: datetime | None = None
    duration_seconds: int | None = None
    result: str | None = None
    approach_notes: str | None = None
    complexity_time: str | None = None
    complexity_space: str | None = None
    llm_review: str | None = None
    comfort_after: int | None = None

    model_config = ConfigDict(from_attributes=True)
